### PR TITLE
add option to pass a function to requestOptions so it can be based dynamically on request's data like filename

### PR DESCRIPTION
**Pull request by borisirota**
Useful for cases like adding the X-SourceMap header which is based on the filename -> [support hidden sourcemaps](https://github.com/getsentry/sentry/issues/4708)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -130,7 +130,8 @@ var config = {
 
 - `deleteAfterCompile`: Boolean determining whether source maps should be deleted on the build server after the webpack compile finishes. Defaults to `false`
 
-- `requestOptions`: Object of options passed through to the underlying `request` call; see the [request library documentation](https://github.com/request/request#requestoptions-callback) for available options.
+- `requestOptions`: Object of options or function returning object of options passed through to the underlying `request` call; see the [request library documentation](https://github.com/request/request#requestoptions-callback) for available options.
+  - If a function is provided, it is given one argument: `req`, the request object passed to the `request library`. (This is useful if you want to add fields dynamically based on request's data like filename, etc.)
```
Comment (schneidmaster) [inline README.md]: Kind of nitpicky but I think this explanation could be a little clearer. Can you consider rephrasing to something like:

> If a function is provided, it is given one argument: `req`, an object of options (including url, auth, and body) that the plugin is sending to the underlying `request` call. (This is useful if you want to configure the request dynamically based on request data such as the filename.)

**Comment by rhys-vdw** [comment]
@schneidmaster I guess there is a use for custom headers after all!

@borisirota seems like a fine idea, but I'm a little concerned that we would be using the same callback for both [`createRelease`](https://github.com/borisirota/webpack-sentry-plugin/blob/981786a0090810dda5752f1b5b53dbc444776820/src/index.js#L156-L170) and [`uploadFile`](https://github.com/borisirota/webpack-sentry-plugin/blob/981786a0090810dda5752f1b5b53dbc444776820/src/index.js#L176-L190). Perhaps it would make sense to have a separate `requestOptions` (function or object) for each of these uses.

Alternatively we could just have a single option for mapping filename to `X-SourceMap` header and specifically add it in `uploadFile`. For example:

```js
  constructor(options) {
    // ...
    this.getFileUploadHeaders = options.pathToSourceMapHeader
      : path => { 'X-SourceMap': options.pathToSourceMapHeader(path) }
      ? () => {}
  }

  uploadFile({ path, name }) {
    const filename = this.filenameTransform(name)
    return request(
      this.combineRequestOptions({
        url: `${this.sentryReleaseUrl()}/${this.releaseVersion}/files/`,
        method: 'POST',
        // not really sure which argument you want here...
        headers: this.getFileUploadHeaders(filename)
        auth: {
          bearer: this.apiKey,
        },
        formData: {
          file: fs.createReadStream(path),
          name: filename,
        },
      }),
    )
  }
```

Full disclosure: I'm not super familiar with this use case or this code base, so let me know if that makes sense. Basically I just want to avoid people getting undefined errors when they add something like this:

```
options {
  requestOptions: req => req.headers['X-SourceMap'] = req.formData.name + '.map'
}
```

**Comment by schneidmaster** [comment]
Ah yeah @rhys-vdw I didn't consider that we'd be using the same function for two different requests. I agree that could cause confusion.

File: README.md (hunk 1 above)
Comment (borisirota) [reply README.md]: Sure

**Comment by borisirota** [comment]
Ohhh first time with the reviews feature so I accidentally submitted a review instead of writing regular comment. Anyway, ignore my comment in the review and this is the original comment:

Hi guys, I think that implementing the general scenario of having both requestOptions is better since it gives the user maximum flexibility without submitting PRs for every specific use case.

In addition I set `uploadFile` headers to be empty object in order to avoid people getting undefined errors as @rhys-vdw mentioned.

I updated the code to reflect this. Let me know what do you think :)

**Comment by schneidmaster** [comment]
@borisirota I like this approach. I think we need one more change, to have both `createReleaseRequestOptions` and `uploadFileRequestOptions` fall back to the `requestOptions` option if provided (and perhaps `console.warn` a deprecation notice in that case, as is done [here](https://github.com/40thieves/webpack-sentry-plugin/blob/master/src/index.js#L20)). Since the currently published version documents `requestOptions`, we can't remove it without breaking backwards compat and I don't want to publish a new major version just for this 🙂

**Comment by borisirota** [comment]
@schneidmaster done 🙂

**Comment by rhys-vdw** [comment]
Looks good @borisirota, thanks for responding so quickly to our 

One last thing to consider is test coverage. Both @schneidmaster and I are new to maintaining this project so I'm not sure what kind of coverage we want to keep. Perhaps I'll allow him to weigh in here, but I think it should be considered before we merge (or at least we can open an issue for adding test coverage in future).

**Comment by schneidmaster** [comment]
@rhys-vdw Yeah I hear you. I think we should open it up as a separate issue. Part of the problem currently is that the tests locally require/run against a real Sentry instance, which makes it a pain to run because you have to set up a test account, an auth token, etc. I would like to rewrite the tests to use mocks for the external service calls instead, which then makes it much easier to expect full test coverage for new code.

Since we don't currently have coverage for `requestOptions`, I think this PR is treading water rather than a step back in terms of missing coverage, so I say we merge it for now and open an issue to implement test mocks/add missing coverage?

**Comment by borisirota** [comment]
thanks guys ! waiting for the next release :)

**Comment by borisirota** [comment]
For some reason, Sentry's source map support through the `X-SourceMap` header isn't working for me.
For future reference I posted [this comment](https://github.com/getsentry/sentry/issues/4708#issuecomment-329227803) on Sentry's repo.

**Pull request merged**